server/src/main/java/org/opensearch/index/store/remote/file/OnDemandBlockIndexInput.java

```diff
@@ -409,4 +416,47 @@ Builder blockSizeShift(int blockSizeShift) {
             return this;
         }
     }
+
+    /**
+     * Simple class to hold the currently open IndexInput backing an instance
+     * of an {@link OnDemandBlockIndexInput}. Lucene may clone one of these
+     * instances, and per the contract[1], the clones will never be closed.
+     * However, closing the instances is critical for our reference counting.
+     * Therefore, we are using the {@link Cleaner} mechanism from the JDK to
+     * close these clones when they become phantom reachable. The clean action
+     * must not hold a reference to the {@link OnDemandBlockIndexInput} itself
+     * (otherwise it would never become phantom reachable!) so we need a wrapper
+     * instance to hold the current underlying IndexInput, while allowing it to
+     * be changed out with different instances as {@link OnDemandBlockIndexInput}
+     * reads through the data.
+     *
+     * [1]: https://github.com/apache/lucene/blob/8340b01c3cc229f33584ce2178b07b8984daa6a9/lucene/core/src/java/org/apache/lucene/store/IndexInput.java#L32-L33
+     */
+    private static class BlockHolder {
+        private IndexInput block;
+
+        private void set(IndexInput block) {
+            if (this.block != null) {
+                throw new IllegalStateException("Previous block was not closed!");
+            }
+            this.block = Objects.requireNonNull(block);
+        }
+
+        private void close() throws IOException {
+            if (block != null) {
+                block.close();
+                block = null;
+            }
+        }
+
+        private void closeQuietly() {
```

Comment:
`IOUtils::closeWhileHandlingException`

Reply:
Not sure to be fair, there is no way to propagate it ... however it seems like Cleaner keeps references on `logger` as well :) What we could do probably is to thrown `UncheckedIOException` instead

Reply:
The Cleaner will silently swallow any unhandled exceptions. I think the logger reference is fine because its a static field and not tied to an instance.

Reply:
👍

Reply:
FYI, the JDK's cleaner implementations do throw UncheckedIOException in this case, no logging, just a note